src/index/DeltaTriples.h

```diff
@@ -0,0 +1,134 @@
+  LocalVocab& localVocab() { return localVocab_; }
+  const LocalVocab& localVocab() const { return localVocab_; }
+
+  // Clear `_triplesAdded` and `_triplesSubtracted` and all associated data
+  // structures.
+  void clear();
+
+  // The number of delta triples added and subtracted.
+  size_t numInserted() const { return triplesInserted_.size(); }
+  size_t numDeleted() const { return triplesDeleted_.size(); }
+
+  // Insert triples.
+  void insertTriples(ad_utility::SharedCancellationHandle cancellationHandle,
+                     std::vector<IdTriple<0>> triples);
+
+  // Delete triples.
+  void deleteTriples(ad_utility::SharedCancellationHandle cancellationHandle,
+                     std::vector<IdTriple<0>> triples);
+
+  // Get `TripleWithPosition` objects for given permutation.
+  LocatedTriplesPerBlock& getLocatedTriplesPerBlock(
+      Permutation::Enum permutation);
+  const LocatedTriplesPerBlock& getLocatedTriplesPerBlock(
+      Permutation::Enum permutation) const;
+
+ private:
+  // Find the position of the given triple in the given permutation and add it
+  // to each of the six `LocatedTriplesPerBlock` maps (one per permutation).
+  // Return the iterators of where it was added (so that we can easily delete it
+  // again from these maps later).
+  std::vector<LocatedTripleHandles> locateAndAddTriples(
+      ad_utility::SharedCancellationHandle cancellationHandle,
+      std::span<const IdTriple<0>> idTriples, bool shouldExist);
+
+  // Erase `LocatedTriple` object from each `LocatedTriplesPerBlock` list. The
+  // argument are iterators for each list, as returned by the method
+  // `locateTripleInAllPermutations` above.
+  //
+  // NOTE: The iterators are invalid afterward. That is OK, as long as we also
+  // delete the respective entry in `triplesInserted_` or `triplesDeleted_`,
+  // which stores these iterators.
+  void eraseTripleInAllPermutations(LocatedTripleHandles& handles);
+};
+
+// DELTA TRIPLES AND THE CACHE
+//
+// For now, our approach only works when the results of index scans are not
+// cached (unless there are no relevant delta triples for a particular scan).
+// There are two ways how this can play out in the future:
+//
+// Either we generally do not cache the results of index scans anymore. This
+// would have various advantages, in particular, joining with something like
+// `rdf:type` would then be possible without storing the whole relation in
+// RAM. However, we need a faster decompression then and maybe a smaller block
+// size (currently 8 MB).
+//
+// Or we add the delta triples when iterating over the cached (uncompressed)
+// result from the index scan. In that case, we would need to (in Step 1 above)
+// store and maintain the positions in those uncompressed index scans. However,
+// this would only work for the results of index scans. For the results of more
+// complex subqueries, it's hard to figure out which delta triples are relevant.
```

Comment:
I asssume that this comment is by @hannahbast , but I think it doesn't go far enough.
Changed `DeltaTriples` invalidate all cache results *that have an IndexScan somewhere in their subtree*, so basically the whole cache is invalidated.
This is something we have to keep in mind when we deal with concurrent updates and queries.